### DISCUSSION TITLE
classes/class_audiostreampolyphonic

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#271] [User-contributed notes]
# classes/class_audiostreampolyphonic

Inherits: AudioStream< Resource< RefCounted< Object AudioStream that lets the user play custom streams at any time from code, simultaneously using a single player. Description: AudioStream that let...

https://docs.godotengine.org/en/stable/classes/class_audiostreampolyphonic.html

**Comment by 1ai13** [comment]
Hello I'm starting with Godot and im working on my second newbie game, after a lot of research, this the solution I implemented for playing multiple audios in parallel:
Declare the AudioPlayer:
```
	private AudioStreamPlayer2D SFXPlayer;
```
Initialize and configure it in you ready or init method (you can use GetNode("XXX") if you have your AudioStreamPlayer in your Scene and not creating it via code):
```
		SFXPlayer = new AudioStreamPlayer2D();
		SFXPlayer.Bus = "SFX";
		//Max number of sounds in parallel
		SFXPlayer.MaxPolyphony = 20;
		//Needed for playing multiple sounds in parallel, PlaybackPolyphonic assigned to  AudioPlayer's Stream
		SFXPlayer.Stream = new AudioStreamPolyphonic();
```
Add it to the scene (skip this step if you initialized it via GetNode<>()):
```
		AddChild(SFXPlayer);
```
Now use the AudioStreamPlayer2D once so u can assign it to the PlaybackPolyphonic and the first sound its not bugged nor triggers NULL exception:
```
		//Playing it without AudioStream for later PolyphonicPlayback assignment without errors
		SFXPlayer.Play();
```
Then in your play method you just need to get your AudioStream resource and play it:
```
		AudioStream sfx = YOUR AUDIO RESOURCE;
			var playback = (AudioStreamPlaybackPolyphonic)SFXPlayer.GetStreamPlayback();
			playback.PlayStream(sfx, 0, volume, 1, 0, "SFX");
```
The params are easy: AudioResource,Offset,Volume,Pitch,PlaybackType (default is 0 you can confiure more with AudioServer.PlaybackType enums) and the Bus.
I hope I helped! I was a bit lost at the beginning with audios, i was creating one AudioStreamPlayer2D Node per each effect i wanted in the Scenes xdd bad approach, now I created an Asset Manager and handled the audio logic there, c u!

**Reply by heyycap** [reply]
@1ai13 Thanks for this! Do you know if there's a way to get the playback position of the individual polyphonic streams? I tried using get_playback_position() on AudioStreamPlaybackPolyphonic, but I get 0. play_stream returns the stream ID, but I haven't figured out how to use that info.

**Reply by FyiurAmron** [reply]
>Now use the AudioStreamPlayer2D once so u can assign it to the PlaybackPolyphonic and the first sound its not bugged nor triggers NULL exception:

why not just use autoplay on the player? Less code, same result.

**Comment by DevSavvySerge** [comment]
How do you play a ne sound w/o interrupting the previous? Stream seems to be a single proiperty.

**Comment by 1ai13** [comment]
The key here is to create a AudioStream2d as usual, but you must set the stream property as AudioStreamPolyphonic() object, if you follow step by step the notes upper in the thread. u should be able to get it:

		SFXPlayer = new AudioStreamPlayer2D();
		SFXPlayer.Bus = "SFX";
		//Max number of sounds in parallel
		SFXPlayer.MaxPolyphony = 20;
		//Needed for playing multiple sounds in parallel, PlaybackPolyphonic assigned to  AudioPlayer's Stream
		SFXPlayer.Stream = new AudioStreamPolyphonic();



		AudioStream sfx = YOUR AUDIO RESOURCE;
			var playback = (AudioStreamPlaybackPolyphonic)SFXPlayer.GetStreamPlayback();
			playback.PlayStream(sfx, 0, volume, 1, 0, "SFX");

**Reply by FyiurAmron** [reply]
`playback.PlayStream(sfx);` is usually enough. You have the offset, volume, pitch, type and bus, none of which is needed explicitly if the player itself is configured with common settings.
